mapper/orm/src/main/java/org/hibernate/search/mapper/orm/massindexing/impl/IdentifierConsumerDocumentProducer.java

```diff
@@ -196,39 +197,85 @@ private void indexAllQueue(SessionImplementor session, List<?> entities) throws
 			return;
 		}
 
-		// TODO use 'SessionWorkExecutor workExecutor = mapping.createSearchManager( session ).createSessionWorkExecutor();' instead
-		PojoWorkPlan workPlan = mapping.createSearchManager( session ).createWorkPlan();
+		// TODO <<monitor>> indexingMonitor.entitiesLoaded( entities.size() );
+		PojoSessionWorkExecutor workExecutor = mapping.createSearchManager( session ).createSessionWorkExecutor();
+		CompletableFuture<?>[] tasks = new CompletableFuture<?>[entities.size()];
 
-		log.tracef( "received a list of objects to index: %s", entities );
-		for ( Object object : entities ) {
-			try {
-				index( workPlan, object );
-				// TODO: implements monitor
-				// monitor.documentsBuilt( 1 );
-			}
-			catch (RuntimeException e) {
-				String errorMsg = log.massIndexerUnableToIndexInstance(
-						object.getClass().getName(),
-						object.toString()
-				);
-
-				// TODO: implements exception handler
-				// errorHandler.handleException( errorMsg, e );
-				// temporary re-throw the exception
-				throw new RuntimeException( errorMsg, e );
+		for ( int i = 0; i < entities.size(); i++ ) {
+			tasks[i] = index( workExecutor, entities.get( i ) );
+		}
+
+		try {
+			CompletableFuture.allOf( tasks ).join();
+		}
+		catch (RuntimeException e) {
+			Integer taskWithExceptionIndex = findAnyCompletedExceptionallyTask( tasks );
+			if ( taskWithExceptionIndex == null ) {
+				throw new RuntimeException( "This row should be never reached, according to CompletableFuture#allOf API", e );
 			}
+
+			// handling error on executing task phase
+			handleException( entities.get( taskWithExceptionIndex ), e );
 		}
 
-		workPlan.execute().get();
+		Integer completedTasks = this.countCompletedTasks( tasks );
+		// TODO <<monitor>> indexingMonitor.documentAdded( completedTasks );
 	}
 
-	private void index(PojoWorkPlan workPlan, Object entity) throws InterruptedException {
+	private CompletableFuture<?> index(PojoSessionWorkExecutor workExecutor, Object entity) throws InterruptedException {
 		// abort if the thread has been interrupted while not in wait(), I/O or similar which themselves would have
 		// raised the InterruptedException
 		if ( Thread.currentThread().isInterrupted() ) {
 			throw new InterruptedException();
 		}
 
-		workPlan.add( entity );
+		// after that a CompletableFuture may not be interrupted anymore, according to the current Java APIs
+		CompletableFuture<?> task = null;
+		try {
+			task = workExecutor.add( entity );
+		}
+		catch (Throwable e) {
+			// handling error on creating task phase
+			handleException( entity, e );
+		}
```

Comment:
Alternatively you can use the pattern we use elsewhere:

```
task = Futures.create( () -> workExecutor.add( entity ) );
```

Then any exception occurring while creating the future will simply result in Futures.create() returning a future that completed exceptionally. Thus you can get rid of the try/catch.

Yes it's a lamba, but in this case, I think it's worth it.